### ISSUE TITLE
HTML source code not showing after Universal toolkit installation in already build angular application

### ISSUE THREAD

**Issue by androidapps10** [maciejtreder/ng-toolkit#743]
## I'm submitting a...

<pre><code>
[ ] Bug report  
[ ] Feature request
[ ] Documentation issue or request
</code></pre>

## In package
<pre><code>
[ ] @ng-toolkit/init
[ ] @ng-toolkit/serverless
[ ] @ng-toolkit/universal
[ ] @ng-toolkit/pwa
[ ] @ng-toolkit/firebug
</code></pre>

## Current behavior
I have a website/application already build in Angular and working fine as per the requirement and the new requirement is to render the HTML for SEO purposes. After installation of Universal Tool it's working like previous, no rendering is performed. Only changes are made in files and source code for Universal toolkit addition.

## Expected behavior
The universal toolkit must show the HTML source code. 

## Environment

<pre><code>
Angular version: 8.3.21 (Global)
Angular version: 7.3.8 (Local)

- Node version: 10.16.3  
- Platform:  Linux
Others:

</code></pre>

**Comment by KingDarBoja** [comment]
Which version of `@ng-toolkit/universal` is being installed at `package.json?`

**Comment by androidapps10** [comment]
> Which version of `@ng-toolkit/universal` is being installed at `package.json?`
-- It's 7.1.2

**Comment by daurea** [comment]
Any insights on why the HTML might not be rendered?

**Comment by KingDarBoja** [comment]
@androidapps10 Can you try with the latest version? It should be `8.0.3`, by running:
```
ng add @ng-toolkit/universal@8.0.3
```

**Comment by androidapps10** [comment]
@KingDarBoja  
I have run the same command: ng add @ng-toolkit/universal@8.0.3
This time no error but still HTML not rendered and the application is still working from built-in JS.

Everything is updated in the package.json file properly.
Check screenshot:  http://prntscr.com/qhwuzm

Please confirm what could be wrong or causing not to show the HTML. It's highly urgent for me to solve the same and show the HTML rendered for SEO purposes.

Ask me if anything needed from my side.

**Comment by KingDarBoja** [comment]
@androidapps10 Which command are you running for testing SSR? It should be `npm run build:ssr` and then `npm run serve:ssr`. It should display you page without problem.

If any doubt, feel free to check my [example repo](https://github.com/KingDarBoja/example-angular-ng-toolkit) which also includes the serverless package.

**Comment by androidapps10** [comment]
@KingDarBoja 

I tried these commands, no luck still many errors and updates needed.
Also FYI, I'm testing the application on my local server. My approach is I work on the local machine, test the application, build the application in production mode and then put on the live server (Live URL).

Please confirm angular universal will work on local with SSR commands or not.

If needed, I'll list the errors list as well for further help.

**Comment by androidapps10** [comment]
@KingDarBoja

After running  'npm run build:ssr' and  'npm run serve:ssr' commands, I'm able to compile the code and the build code is there in /dist folder 
Screesnhot:   http://prntscr.com/qir6xl

Now the latest error i'm facing is in sist/server.js file window variable issue.

dist/server.js:135865
var $ = window["jQuery"];
        ^
ReferenceError: window is not defined

There were some other errors as well which are solved but still listing [See Here Error List](http://webappsitesdemo.com//angular_universal/angular_universal_errors_list.txt). Please check and confirm with the next steps to follow up.

Looking forward to hearing back from you on same soon.

**Comment by androidapps10** [comment]
@KingDarBoja 

Now the window error has been resolved by using this:

const domino = require('domino');
const fs = require('fs');
const path = require('path');
const template = fs.readFileSync(path.join(__dirname, '.', 'dist', 'index.html')).toString();
const win = domino.createWindow(template);
global['window'] = win;
global['document'] = win.document;

The new error is as follows: 


> node local.js

fs.js:114
    throw err;
    ^

Error: ENOENT: no such file or directory, open '/dist/index.html'
    at Object.openSync (fs.js:443:3)
    at Object.readFileSync (fs.js:343:35)
    at Module.<anonymous> (/var/www/html/Digital-test/dist/server.js:117:19)
    at Module.<anonymous> (/var/www/html/Digital-test/dist/server.js:147:30)
    at __webpack_require__ (/var/www/html/Digital-test/dist/server.js:21:30)
    at /var/www/html/Digital-test/dist/server.js:85:18
    at Object.<anonymous> (/var/www/html/Digital-test/dist/server.js:88:10)
    at Module._compile (internal/modules/cjs/loader.js:778:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:789:10)
    at Module.load (internal/modules/cjs/loader.js:653:32)

**Comment by KingDarBoja** [comment]
@androidapps10 Did you checked the content of `local.js` file in order to be sure it is pointing to the correct path? Looks like it is looking for `dist/index.html` but it doesn't exist. It should be `/dist/server` the folder to look at.

**Comment by androidapps10** [comment]
@KingDarBoja 
Yes I have checked local.js file and the path set is for /dist/server
const server = require('./dist/server');

Looks like it is looking for dist/index.html but it doesn't exist. It should be /dist/server the folder to look at.
->In response to this, I have noticed that after building the code with 'npm run build:ssr' command the index.html is under the directory 'dist/browser' instead of 'dist/server'.

FYI, I have changed the template path
const template = fs.readFileSync(path.join(__dirname, '.', 'dist', 'index.html')).toString(); 
 to this one:
const template = fs.readFileSync(path.join(__dirname, join(DIST_FOLDER, '/index.html'))).toString(); 
 in the server.js file  DIST_FOLDER  path is as follows:
const DIST_FOLDER = join(process.cwd(), 'dist/browser');

 So basically the application is calling index.html from dist/browser instead of dist/server.

Now, I have copied the build code on my system as local but still not able to get the HTML rendered.

Assert me to follow the next steps OR ask me if any further information needed from my side.

**Comment by KingDarBoja** [comment]
> In response to this, I have noticed that after building the code with 'npm run build:ssr' command the index.html is under the directory 'dist/browser' instead of 'dist/server'.

This is the correct behaviour, the `index.html` is located at the `dist/browser` folder and the `local.js` file is looking at `dist/server` because it contains the server file which was generated from `server.ts`.

The problem is indeed inside at `server.ts` file as it cannot find the index one. Have you tried copying the content of my sample repo?

These files:
- [server.ts](https://github.com/KingDarBoja/example-angular-ng-toolkit/blob/master/server.ts)
- [main.server.ts](https://github.com/KingDarBoja/example-angular-ng-toolkit/blob/master/src/main.server.ts)
- [main.ts](https://github.com/KingDarBoja/example-angular-ng-toolkit/blob/master/src/main.ts)

As they are the ones that are being used at ssr. Also, run (or replace) `node dist/server` instead of the `serve:ssr` command as I don't use anymore `local.js` on my sample repo.

Also, be sure to update your app to Angular 8, as I didn't noticed you are using an old version (@angular/core is 7.x.x for example...).

Cheers!

**Comment by androidapps10** [comment]
@KingDarBoja
The problem is indeed inside at server.ts file as it cannot find the index one.
-> In response to this, not sure what exact issue is there.

Have you tried copying the content of my sample repo? These files:
server.ts
main.server.ts
main.ts

-> In response to this, yes I checked and cloned the same on my local, the application worked but without HTML rendering.

Also, run (or replace) node dist/server instead of the serve:ssr
->Tried this as well.

Also, be sure to update your app to Angular 8, as I didn't notice you are using an old version (@angular/core is 7.x.x for example...).
-> Yes, I have updated the APP to Angular 8 now. 

After following all you above shared steps and commands the latest result is as follows and Application is failed to compile: 
1. in local.js file 
const server = require('./dist/browser'); has been changed earlier it was
 const server = require('./dist/server');

2. While building the application:
node local.js

internal/modules/cjs/loader.js:638
    throw err;
    ^
3. I should update my local.js or server.js because the path to pick the HTML is not the correct one ?

Error: Cannot find module 'require("./server/main")'

Now it's looking for server folder instead of browser in /dist/

Looking forward to hearing back from you on the same soon.

**Comment by KingDarBoja** [comment]
Let's make another try, do you `webpack.config.js` file has this line:
```js
externals: {
	'./dist/server/main': 'require("./server/main")'
},
```

If it does, please comment it and try again.

If that doesn't work, better share some repository to look at as I ran out of ideas.

**Comment by androidapps10** [comment]
@KingDarBoja 

I have checked my webpack.server.config.js file and commented on the lines of code you have shared.
Now the new error appears as follow, where  '**Frontend**' is the name of my working directory

Listening on: http://localhost:4000
Error: No NgModule metadata found for '[object Object]'.
    at NgModuleResolver.module.exports.NgModuleResolver.resolve (/var/www/html/Frontend/dist/server.js:69141:23)
    at CompileMetadataResolver.module.exports.CompileMetadataResolver.getNgModuleMetadata (/var/www/html/Frontend/dist/server.js:68248:43)
    at JitCompiler.module.exports.JitCompiler._loadModules (/var/www/html/Frontend/dist/server.js:74429:51)
    at JitCompiler.module.exports.JitCompiler._compileModuleAndComponents (/var/www/html/Frontend/dist/server.js:74410:36)
    at JitCompiler.module.exports.JitCompiler.compileModuleAsync (/var/www/html/Frontend/dist/server.js:74370:37)
    at CompilerImpl.module.exports.CompilerImpl.compileModuleAsync (/var/www/html/Frontend/dist/server.js:94195:31)
    at /var/www/html/Frontend/dist/server.js:46705:22
    at new ZoneAwarePromise (/var/www/html/Frontend/dist/server.js:2225:29)
    at getFactory (/var/www/html/Frontend/dist/server.js:46691:12)
    at View.engine (/var/www/html/Frontend/dist/server.js:46670:13)

**Comment by KingDarBoja** [comment]
@androidapps10 Those are angular related problems. Please take a look at [this](https://stackoverflow.com/questions/50477712/error-no-ngmodule-metadata-found-for-object-object) and [this](https://stackoverflow.com/questions/54251353/angular-universal-getting-the-no-ngmodule-metadata-found-for-object-object) StackOverflow questions.

**Comment by androidapps10** [comment]
@KingDarBoja 
Thanks for your reply, I have checked both the links:

-> About the bundleDependencies, I don't have this term defined in my angular.json file.

-> About the second point, I have added :
import 'reflect-metadata';
Which was earlier missing in the file then clear cache, build again and then serve again.
The issue remains the same.

Anything else that might be missing or I should update or check.

Looking forward to hearing back from you on the same soon.

**Comment by androidapps10** [comment]
@KingDarBoja
I have managed to remove all the errors related to NgModule and Buffer deprecation issue after the ssr build and server.

Now the last issue pending is when I'm running the npm run serve:ssr, it listens to server request but throws errors line like this: http://prntscr.com/qnklcl in the terminal and
Is there any way to get out of this issue, because it's not even referencing any line number.
Everything is up-to-date with the latest dependencies now for my application.

Thanks in advance.

**Comment by KingDarBoja** [comment]
Sorry for the delay, been busy with job stuff but glad you solve most issues on your project.

About the error, hard to know what is causing the issue if it doesn't throw any stack trace. What you can do is try-catch the exception on the server.ts file and log it.

**Comment by androidapps10** [comment]
@KingDarBoja 
I have checked the logs and it's throwing me the ng-bootstrap error: No matching version found for @ng-bootstrap/ng-bootstrap@4.3.1

Though I have cross-checked my dependencies and version 4.3.1 is installed. Even its mentioned in the official documentation that it's available for Angular 8:
https://ng-bootstrap.github.io/#/getting-started

Will you please suggest fixing the same or any other alternative.

**Comment by KingDarBoja** [comment]
For some reason, the `@ng-bootstrap/ng-bootstrap@4.3.1` is not available in npm as seen on the [versions tab](https://www.npmjs.com/package/@ng-bootstrap/ng-bootstrap?activeTab=versions).

Try installing the `4.2.2` one.

**Comment by androidapps10** [comment]
@KingDarBoja 
Tried installing 4.2.2 and it comes with a lot of other errors regarding peer dependencies.
I have added all other dependencies as well now.

What kind of issue related to project config, will you please clear me more on same?
